src/lib/pubkey/hss_lms/lms.cpp

```diff
@@ -0,0 +1,428 @@
+   const auto hash = HashFunction::create_or_throw(m_hash_name);
+   m_m = hash->output_length();
+}
+
+LMS_PublicKey LMS_PrivateKey::sign_and_get_pk(StrongSpan<LMS_Signature_Bytes> out_sig,
+                                              LMS_Tree_Node_Idx q,
+                                              const LMS_Message& msg) const {
+   // Pre-alloc space for the signature
+   BOTAN_ARG_CHECK(out_sig.size() == LMS_Signature::size(lms_params(), lmots_params()), "Invalid output buffer size");
+
+   BufferStuffer sig_stuffer(out_sig);
+   sig_stuffer.append(store_be(q));
+   const LMOTS_Private_Key lmots_sk(lmots_params(), identifier(), q, seed());
+   lmots_sk.sign(sig_stuffer.next<LMOTS_Signature_Bytes>(LMOTS_Signature::size(lmots_params())), msg);
+   sig_stuffer.append(store_be(lms_params().algorithm_type()));
+   const auto auth_path_buffer = sig_stuffer.next<LMS_AuthenticationPath>(lms_params().m() * lms_params().h());
+
+   BOTAN_ASSERT_NOMSG(sig_stuffer.full());
+
+   TreeAddress lms_tree_address(lms_params().h());
+   LMS_Tree_Node pk_buffer(lms_params().m());
+   lms_treehash(StrongSpan<LMS_Tree_Node>(pk_buffer.get()), auth_path_buffer, q, *this);
+
+   return LMS_PublicKey(lms_params(), lmots_params(), identifier(), std::move(pk_buffer));
+}
+
+LMS_PublicKey LMS_PublicKey::from_bytes_or_throw(BufferSlicer& slicer) {
+   size_t total_remaining_bytes = slicer.remaining();
+   // Alg. 6. 1. (4 bytes are sufficient until the next check)
+   if(total_remaining_bytes < sizeof(LMS_Algorithm_Type)) {
+      throw Decoding_Error("Too few bytes while parsing LMS public key.");
+   }
+   // Alg. 6. 2.a.
+   auto lms_type = load_be<LMS_Algorithm_Type>(slicer.take<sizeof(LMS_Algorithm_Type)>());
+   // Alg. 6. 2.c.
+   auto lms_params = LMS_Params::create_or_throw(lms_type);
+   // Alg. 6. 2.d.
+   if(total_remaining_bytes < LMS_PublicKey::size(lms_params)) {
+      throw Decoding_Error("Too few bytes while parsing LMS public key.");
+   }
+   // Alg. 6. 2.b.
+   auto lmots_type = load_be<LMOTS_Algorithm_Type>(slicer.take<sizeof(LMOTS_Algorithm_Type)>());
+   auto lmots_params = LMOTS_Params::create_or_throw(lmots_type);
+
+   if(lms_params.hash_name() != lmots_params.hash_name()) {
+      throw Decoding_Error("No support for HSS-LMS instances with multiple hash functions.");
+   }
+
+   // Alg. 6. 2.e.
+   auto I = slicer.copy<LMS_Identifier>(LMS_IDENTIFIER_LEN);
+   // Alg. 6. 2.f.
+   auto lms_root = slicer.copy<LMS_Tree_Node>(lms_params.m());
+
+   return LMS_PublicKey(std::move(lms_params), std::move(lmots_params), std::move(I), std::move(lms_root));
+}
+
+std::vector<uint8_t> LMS_PublicKey::to_bytes() const {
+   std::vector<uint8_t> bytes(size(lms_params()));
+   BufferStuffer stuffer(bytes);
+
+   stuffer.append(store_be(lms_params().algorithm_type()));
+   stuffer.append(store_be(lmots_params().algorithm_type()));
+   stuffer.append(identifier());
+   stuffer.append(m_lms_root);
+   BOTAN_ASSERT_NOMSG(stuffer.full());
+
+   return bytes;
```

Comment:
Technically, we could now formulate that with `concat()`:

```suggestion
   // clang-format off
   return concat<std::vector<uint8_t>>(
      store_be(lms_params().algorithm_type()),
      store_be(lmots_params().algorithm_type()),
      identifier(),
      m_lms_root);
   // clang-format on
```

... your call on what is easier to read.

Reply:
I like your version better!